quil-rs/src/program/mod.rs

```diff
@@ -379,6 +381,76 @@ impl Program {
         Ok(expanded_program)
     }
 
+    /// Return a copy of the [`Program`] wrapped in a loop that repeats `iterations` times.
+    ///
+    /// The loop is constructed by wrapping the body of the program in classical Quil instructions.
+    /// The given `loop_count_reference` must refer to an INTEGER memory region. The value at the
+    /// reference given will be set to `iterations` and decremented in the loop. The loop will
+    /// terminate when the reference reaches 0. For this reason your program should not itself
+    /// modify the value at the reference unless you intend to modify the remaining number of
+    /// iterations (i.e. to break the loop).
+    ///
+    /// The given `start_target` and `end_target` will be used as the entry and exit points for the
+    /// loop, respectively. You should provide unique [`Target`]s that won't be used elsewhere in
+    /// the program.
+    ///
+    /// If `iterations` is 0, then a copy of the program is returned without any changes.
+    pub fn wrap_in_loop(
+        &self,
+        loop_count_reference: MemoryReference,
+        start_target: Target,
+        end_target: Target,
+        iterations: u32,
+    ) -> Self {
+        if iterations == 0 {
+            return self.clone();
+        }
+
+        let mut looped_program = self.clone_without_body_instructions();
+
+        looped_program.add_instructions(
+            vec![
+                Instruction::Declaration(Declaration {
+                    name: loop_count_reference.name.clone(),
+                    size: Vector {
+                        data_type: ScalarType::Integer,
+                        length: 1,
+                    },
+                    sharing: None,
+                }),
+                Instruction::Move(Move {
+                    destination: loop_count_reference.clone(),
+                    source: ArithmeticOperand::LiteralInteger(iterations.into()),
+                }),
+                Instruction::Label(Label {
+                    target: start_target.clone(),
+                }),
+            ]
+            .into_iter()
+            .chain(self.body_instructions().cloned())
+            .chain(vec![
+                Instruction::Arithmetic(Arithmetic {
+                    operator: ArithmeticOperator::Subtract,
+                    destination: ArithmeticOperand::MemoryReference(MemoryReference {
+                        name: "__loop-count".to_string(),
```

Comment:
This could be a problem if the user wants to use multiple loops (e.g. inner and outer)

Reply:
Shouldn't this be `loop_count_reference.name`?

Reply:
Yup, that explains the snapshot issue too, good catch.